14-Filter-contains/doesn't-contain-should-be-case-insensitive issues solved

Nitinyadav237 opened the pull request:
**Description:**
I have added a step to convert both the selected column and the filter value string to lowercase before comparison. This ensures that the comparison is performed without worrying about the case of the letters.

Fix: #14


https://github.com/pretzelai/pretzelai/assets/113238477/b83b4d8f-811d-434b-a675-a6976b6a1b4f

ramonverse commented:
Hi @Nitinyadav237 , thanks for this PR. Please remove the changes to package.json, package-lock and worker.ts

Nitinyadav237 commented:
i have removed the changes changes to package.json, package-lock and worker.ts

in worker.ts after adding/public/  
![image](https://github.com/pretzelai/pretzelai/assets/113238477/e792e7ba-27fa-4384-a5c6-2917e6d902c9)

Nitinyadav237 commented:
https://github.com/pretzelai/pretzelai/assets/113238477/0b1deead-e527-47a3-bca6-1634dc022ad6

ramonverse commented on src/lib/utils.ts:
```diff
@@ -71,10 +71,13 @@ export const columnErrorQueryBuilder = (columns: string[]) => {
-  return filter ? `filter ${filter}` : ""
+  if (filter && filter.includes("text.lower") && filter.includes("not")) {
+    // If the filter includes both "text.lower" and "not"
+    return `filter !${filter.replace("not","")}`;
```
this is introducing a bug, it is the wrong place to add this transformation, should be handled in FilterBlock

Nitinyadav237 replied:
initially filterQueryBuilder was like this 
export const filterQueryBuilder = (filter: string) => {
    return filter ? `filter ${filter}` : "";
}

to avoid case sensitivity i decided to lowercase both column value and valuestr 

so i have used inbuilt method text.lower to convert column value to lowercase 
![image](https://github.com/pretzelai/pretzelai/assets/113238477/e6375b0c-b5dd-4ba9-b12a-f8f7f9e25bc8)


 operatorStr = "| text.lower | text.contains"
        valueStr=valueStr.toLowerCase()

but the issue came with  doesn't contain final sql query  generated was "SELECT
  *
FROM
  "InputTable"
WHERE
  LOWER(NOT "Name") LIKE CONCAT('%', 'acode', '%')
LIMIT
  10000
"
to overcome i decided to change the query only for "not contains"

 if (filter && filter.includes("text.lower") && filter.includes("not")) {
    // If the filter includes both "text.lower" and "not"
    return `filter !${filter.replace("not","")}`;
  }
this condition check whether filter includes "textlower" and includes "not" i have added this condition  `filter !${filter.replace("not","")}`;
this resulted in correct sql query 
"SELECT
  *
FROM
  "InputTable"
WHERE
  NOT LOWER("Name") LIKE CONCAT('%', 'acode', '%')
LIMIT
  10000
"

and the issue solved finally

ramonverse reviewed (changes requested):
Hi @Nitinyadav237 , you are negating the entire filter by introducing the transformation in utils. Utils should be left unchanged and handle the negative case in the filterblock filter builder.
I acknowledge filterblock is a very messy component that could use some refactoring. If you are down for it, we could use some unit testing for filterblock to make sure we don't break stuff (let me know if you want to take care of testing and I'll explain what I had in mind).

An example of the bug I mentioned (you are negating all the filters so Equals become effectively not equals)
<img width="1561" alt="image" src="https://github.com/pretzelai/pretzelai/assets/121360087/5f9ce5b7-f68c-4808-abb3-ac690dd0bdd3">

Nitinyadav237 commented:
![image](https://github.com/pretzelai/pretzelai/assets/113238477/68ceac73-b1b2-4264-ba83-c61a67819da0)

this is the final sql query generated
"SELECT
  *
FROM
  "InputTable"
WHERE
  "Name" = 'trix'
  AND NOT "Name" LIKE CONCAT('%', 'AcoDE', '%')
  OR "Full Name" = 'base'
LIMIT
  10000
"

Nitinyadav237 commented:
![image](https://github.com/pretzelai/pretzelai/assets/113238477/cb42d164-a6f5-4f13-860e-d2b059235837)

This is the final sql query generated 
"SELECT
  *
FROM
  "InputTable"
WHERE
  "Name" = 'trix'
  AND NOT "Name" LIKE CONCAT('%', 'AcoDE', '%')
  OR "Full Name" = 'base'
LIMIT
  10000
"

ramonverse commented:
> is it okay or do i need to change it ?

Hi @Nitinyadav237 , can you make the changes in FilterBlock? You should construct the query there not in utils. Utils should remain unchanged.

Nitinyadav237 commented:
Implemented case-insensitivity logic within the filter block for 'contains/doesn't contain' conditions

![image](https://github.com/pretzelai/pretzelai/assets/113238477/f1be5daa-d317-4881-8200-bbfb92a85d94)

final sql query : "SELECT
  *
FROM
  "InputTable"
WHERE
  LOWER("Name") LIKE CONCAT('%', 'trix', '%')
  AND NOT LOWER("Name") LIKE CONCAT('%', 'react', '%')
LIMIT
  10000
"

Nitinyadav237 commented:
@ramonverse can you review it?

ramonverse commented:
Hi @Nitinyadav237 , I didn't have time to go deep into this but it seems like this logic shouldn't be in a useEffect. Can you do the logic inside the constructSQL function? Add the "not" cases as extra `else if` statements if you have to. If that is not possible happy to learn why.

<img width="702" alt="image" src="https://github.com/pretzelai/pretzelai/assets/121360087/7dc8f969-947c-4646-ab3d-750b1d6fa85f">

Nitinyadav237 commented:
@ramonverse  i have remove the logic from the useeffect . 

const applyFilter = () => {
  const filter = filterQueryBuilder(constructSQL(rootFilterGroup));
  let q = mergeQueries(prevQuery, filter);

  if (filter.includes("text.lower") && filter.includes("not ")) {
    const parts = filter.split(/\s*(\|\||&&)\s*/);
    const modifiedFilter = parts
      .map((part) => {
        if (part.includes("text.lower") && part.includes("not")) {
          const replacedNeg = `!${part.replace("not ", "")}`;
          return replacedNeg.replace("!filter", "filter !");
        } else {
          return part; // Return the original part if not modified
        }
      })
      .join(" ");
    q = mergeQueries(prevQuery, modifiedFilter);
  }

  updateQuery(q);
};

useEffect(() => {
  applyFilter();
}, [rootFilterGroup, prevQuery]);

Nitinyadav237 commented:
1. **Case Sensitivity Handling**:
              Lowercased both column values and search terms (valueStr) using text.lower.
2. **Issue with "Doesn't Contain" Operator:**
             Encountered problem when text.lower treated "not" as a parameter, affecting SQL query generation.
3. **Solution Approach:**
             Implemented applyFilter function to manage query generation based on filter criteria.
4. **Handling "Doesn't Contain" Operator**:
              Within applyFilter, checked if query contained both "text.lower" and "not"
5. **Modification Strategy**:
           Replaced "not " with empty string and added "!" before it to fix "doesn't contain" issue.
6. **Finalization and Update:**
            Merged modified filter with previous query and updated it using updateQuery.
By taking these steps, we ensured correct behavior in SQL query generation, particularly addressing the "doesn't contain" operator issue.

ramonverse commented:
@Nitinyadav237 wouldn't this work better? 
<img width="978" alt="image" src="https://github.com/pretzelai/pretzelai/assets/121360087/4273ea97-1671-4e05-a266-4bc9db40abc3">

Nitinyadav237 commented:
> @Nitinyadav237 wouldn't this work better? <img alt="image" width="978" src="https://private-user-images.githubusercontent.com/121360087/318645340-4273ea97-1671-4e05-a266-4bc9db40abc3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwMzQzMzcsIm5iZiI6MTcxMjAzNDAzNywicGF0aCI6Ii8xMjEzNjAwODcvMzE4NjQ1MzQwLTQyNzNlYTk3LTE2NzEtNGUwNS1hMjY2LTRiYzlkYjQwYWJjMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwMlQwNTAwMzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYzE1ZTNjYTkzYzE0ZWQzOTkwYmVjNDRkYjQ2ZTY5NzYzNWU1ZmVkYTBlNTY1MmI0NTFjODZlNDVlY2QxYjU5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.s1PoQS1Grw9RQYzOkc40O8f-Bvy31C-3vkg1KWf14Zw">

SO you are only performing lowercase in the "contain case " and not in  "doesn't contain case"

Nitinyadav237 commented:
@ramonverse can i close this pull request

khareyash05 commented:
> @ramonverse can i close this pull request?

Hey @Nitinyadav237 I would like to collaborate on this PR.

khareyash05 commented:
Hey are you working on it? @Nitinyadav237

Nitinyadav237 commented:
> Hey are you working on it? @Nitinyadav237

issued is solved by maintainer itself

ramonverse commented:
This PR still needs to be solved, we just need to put the code in the right place. Yes, it should work for contain & not contain, I added the screenshot to point you in the right direction. @khareyash05 feel free to push the final code to this PR

Nitinyadav237 commented:
@ramonverse 

![image](https://github.com/pretzelai/pretzelai/assets/113238477/e8dcd627-bc12-4e70-a464-713ce5dae61f)

I have question for you why there is no need of performing lowercase column name in case of "doesn't contain" but we are lowercasing for "contain" .
This is the final sql query generated 
"SELECT
  *
FROM
  "InputTable"
WHERE
  LOWER("Name") LIKE CONCAT('%', 'atom', '%')
  AND NOT "Name" LIKE CONCAT('%', 'keyboard', '%')
LIMIT
  10000
"

ramonverse commented:
Hey @Nitinyadav237 it should definitely lowercase it for "doesn't contain" too as indicated in the title of the PR. I was just pointing you in the right direction but as you correctly spotted there is a bug in my snippet because it is not lowercased. If you fix that I think the PR should be good to go!

Nitinyadav237 commented:
@ramonverse now it is generating sql query as expected
"SELECT
  *
FROM
  "InputTable"
WHERE
  LOWER("Name") LIKE CONCAT('%', 'atom', '%')
  AND NOT LOWER("Name") LIKE CONCAT('%', 'keyboard', '%')
LIMIT
  10000
"
